### ISSUE TITLE
refactoring compute_stats_r & route_integrity

### ISSUE THREAD

**Issue by yakra** [yakra/DataProcessing#236]
### `compute_stats_r`
is where it is simply because it was split off from `compute_stats_t`, and related items were left grouped together. Plus, what eventually became `RteIntThread` [didn't exist yet](https://github.com/yakra/DataProcessing/commit/b07f73b1a6531aad7ca8e0f9cc5ce81d2be0a45e).
The datachecks were added in because it was a handy place to put them, in a multi-threaded job iterating thru systems & routes, *after* .list files were processed. The original plan was to have what became ABBREV_NO_CITY [do stuff based on whether a .list name was in use](https://github.com/TravelMapping/DataProcessing/issues/303#issuecomment-611341196). So, [ABBREV_AS_BANNER](https://github.com/yakra/DataProcessing/blame/7cf59eb5a4a1f21600a42b23d7617fad4836ac95/siteupdate/cplusplus/classes/Route/compute_stats_r.cpp#L59-L64) was added, with the idea that when it eventually came time to implement ABBREV_NO_CITY, I'd just slap an `else` after the whole shebang and go to town. By the time it was [implemented](https://github.com/yakra/DataProcessing/commit/94560f1848f2442536cb4d94ca90303b8eae1dd1#diff-cfffce5fd934b05229e6ec96117555b4725f1bd31d1e84b5d90c64bdab6c1406), I'd decided against checking .list names in use because [There's no way to automatically know whether the error is due to transposed data as opposed to a missing city or extraneous abbrev.](https://github.com/TravelMapping/DataProcessing/issues/303#issuecomment-963316444)
**TL;DR**, *this doesn't have to be where it is,* after .list processing. It can be refactored into `HighwaySystem::route_integrity`.
With each thread now doing other stuff the majority of the time, that could mean less competition for the regional mutexes, and a win for parallelism. *(OTOH, doing more stuff at once could mean increased cache misses and thus decreased performance.)*
* [ ] This is as simple as putting a call to `compute_stats_r` into `route_integrity`, where it already iterates https://github.com/yakra/DataProcessing/blob/16721240fb5c3e117b1701b62bfcda6a10fb0c61/siteupdate/cplusplus/classes/HighwaySystem/route_integrity.cpp#L10
  We could then get rid of `CompStatsRThread` and the code in `main` that processes it.

---

### `route_integrity`
is where it is because the original switch to efficient-but-destructive one-time case-smashing on AltLabels required it to happen after NMP processing, in order to preserve case in nearmisspoints.log & nmpfps.log and allow FP entries to match without big changes to nmpfps.log. [AltLabels have since been removed from these logs entirely](https://github.com/TravelMapping/DataProcessing/pull/430), meaning we can now smash case anywhere.
The errorcheck for Routes without a ConnectedRoute was lumped in to take advantage of the existing iteration through systems & routes.
* **Near-term:**
  * [ ] Moving label hash creation into `Route::read_wpt` (where the UAL set was originally populated) would help out during initial "dry runs" where `hwy_data/` isn't cached, by performing this task while waiting for disk access.
    This could also ameliorate the potential cache miss effects mentioned above.
  * [ ] Moving root hash & list hash creation out of the `Route` ctor and into `route_integrity` (effectively putting it back in its old location after concurrency detection) would move it from a single to multi-threaded part of the program.     (Cache misses back, potentially?)
    * Need mutex(es). High potential for contention? Not so bad if interspersed with other tasks? C++17 `shared_mutex`, maybe?
* **Long-term:**
  * [ ] Threaded `HighwaySystem` construction, exact mechanics TBD.
    * Could have one function create the `HighwaySystem` objects, followed by a threaded `read_csv` function, similar to what's done with .wpt files now.
    * Or create a vector of `systems.csv` lines & have a threaded funtion iterate thru that & call the ctor.
  * [ ] Move the connectivity checks into the `HighwaySystem` ctor & get rid of `route_integrity` entirely.
  * [ ] Move the call to `read_wpt` into the `HighwaySystem` ctor. We'd then be up 1 thread (system construction) & down 3 (ReadWptThread, RteIntThread, CompStatsRThread), and doing a lot more stuff while waiting for disk access.

---

### Low priority
* `Route` branch, [DUPLICATE_LABEL, pointsinuse & unusedaltlabels bugfixes](https://github.com/TravelMapping/DataProcessing/issues/412) ought to happen first.

**Comment by yakra** [comment]
Solving the [region.php rankings bug](https://forum.travelmapping.net/index.php?topic=5379) while maintaining good performance means reunifying `CompStatsRThread` + `CompStatsTThread`& iterating by region.
Nothing else iterates by region, so the new combined `CompStatsThread` stays; there's nothing else to refactor it into.

I'll close this issue and replace it with another one that reflects that reality soon.

**Comment by yakra** [comment]
OTOH, the LABEL_SELFREF datacheck requires that all .wpts are read, that all colocated points are detected. It has to happen after ReadWptThread.
The only other other options are NMPSearchThread & NMPMergedThread, which just don't make sense from an organizational standpoint. I think I'll leave this well enough alone.
